include/seqan3/io/alignment_file/format_sam.hpp

```diff
@@ -412,131 +1237,135 @@ class alignment_file_format_sam
         }
     }
 
-    /*!\brief Writes the SAM header_ptr.
+    /*!\brief Writes the SAM header.
      * \tparam stream_t   The stream type.
      *
      * \param[in,out] stream  The stream to print to.
      * \param[in]     options The options to alter printing.
-     * \param[in]     header_ptr  The header_ptr (as a pointer) to print.
+     * \param[in]     header  The header to print.
      *
-     * \throws seqan3::format_error if the header_ptr object contains the wrong
+     * \throws seqan3::format_error if the header object contains the wrong
      *         information or the contents are ill-formed.
      *
      * \details
      *
-     * Before writing the header_ptr, the contents are checked for correctness
+     * Before writing the header, the contents are checked for correctness
      * according to the rules of the official
      * [SAM format specifications](https://samtools.github.io/hts-specs/SAMv1.pdf).
      */
-    template <typename stream_t>
+    template <typename stream_t, typename ref_ids_type>
     void write_header(stream_t & stream,
                       alignment_file_output_options const & options,
-                      std::unique_ptr<alignment_file_header> & header_ptr)
+                      alignment_file_header<ref_ids_type> & header)
     {
-        if (header_ptr != nullptr)
-        {
-            // -----------------------------------------------------------------
-            // Check Header
-            // -----------------------------------------------------------------
+        // -----------------------------------------------------------------
+        // Check Header
+        // -----------------------------------------------------------------
+
+        // (@HD) Check header line
+        // The format version string will be taken from the local member variable
+        if (!header.sorting.empty() &&
+            !(header.sorting == "unknown"   ||
+              header.sorting == "unsorted"  ||
+              header.sorting == "queryname" ||
+              header.sorting == "coordinate" ))
+            throw format_error{"SAM format error: The header.sorting member must be "
+                               "one of [unknown, unsorted, queryname, coordinate]."};
+
+        if (!header.grouping.empty() &&
+            !(header.grouping == "none"   ||
+              header.grouping == "query"  ||
+              header.grouping == "reference"))
+            throw format_error{"SAM format error: The header.grouping member must be "
+                               "one of [none, query, reference]."};
```

Comment:
Shouldn't these header members be enums maybe?

Reply:
I thought to reduce the data types in the header because there would be enum types for grouping, subgrouping and sorting...

Reply:
Hm, I but enums can be well documented and you don't need to enforce values vai string comparison... I tihnk this is exactly what enums are for!

But we can make a card to change this later, I assume the header will get more members over time anyway.